groth16/src/proof.rs

```diff
@@ -0,0 +1,186 @@
+#![allow(non_snake_case)]
+
+use std::ops::{AddAssign, Mul};
+use ark_ec::pairing::Pairing;
+use ark_ff::{FftField, PrimeField};
+use ark_groth16::Proof;
+
+fn calculate_A<
+    F: FftField + PrimeField + Into<u64>,
+    E: Pairing<G1Affine = F>,
+>(
+    L: F,
+    N: F,
+    r: F,
+    S: Vec<F>,
+    a: Vec<F>,
+) -> E::G1Affine {
+    // Start out by calculating the product of S_i^a_i
+    let mut prod = F::one();
+    for i in 0..S.len() {
+        prod *= S[i].pow(&[a[i].into()]);
+    }
+
+    // Calculate L * (N)^r
+    let lhs = L * N.pow(&[r.into()]);
+
+    // Finally, multiply lhs by prod to find A
+    lhs * prod
+}
+
+fn calculate_B<
+    F1: FftField + PrimeField + Into<u64>,
+    F2: FftField + PrimeField + Into<u64>,
+    E: Pairing<G2Affine = F2>,
+>(
+    Z: F2,
+    K: F2,
+    s: F2,
+    V: Vec<F2>,
+    a: Vec<F1>,
+) -> E::G2Affine {
+    // Start out by calculating the product of V_i^a_i
+    let mut prod = F2::one();
+    for i in 0..V.len() {
+        prod *= V[i].pow(&[a[i].into()]);
+    }
+
+    // Calculate Z * (K)^s
+    let lhs = Z * K.pow(&[s.into()]);
+
+    // Finally, multiply lhs by prod to find B
+    lhs * prod
+}
+
+fn calculate_h_of_x<
```

Comment:
As mentioned this is already covered by `ext_wit`. Non need to be added this IMO